cpp/src/parquet/stream_reader.cc

```diff
@@ -0,0 +1,224 @@
+  read_<DoubleReader>(v);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(char& v) {
+  checkColumn_(Type::FIXED_LEN_BYTE_ARRAY, ConvertedType::NONE, 1);
+  FixedLenByteArray flba;
+  read_(flba);
+  v = static_cast<char>(flba.ptr[0]);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::string& v) {
+  checkColumn_(Type::BYTE_ARRAY, ConvertedType::UTF8);
+  ByteArray ba;
+  read_(ba);
+  v = std::string(reinterpret_cast<const char*>(ba.ptr), ba.len);
+  return *this;
+}
+
+void StreamReader::readFixedLength_(char* ptr, const int len) {
+  checkColumn_(Type::FIXED_LEN_BYTE_ARRAY, ConvertedType::NONE, len);
+  FixedLenByteArray flba;
+  read_(flba);
+  std::memcpy(ptr, flba.ptr, len);
+}
+
+void StreamReader::read_(ByteArray& v) {
+  const auto& node{nodes_[column_index_]};
+  auto* const reader{
+      static_cast<ByteArrayReader*>(column_readers_[column_index_++].get())};
+  std::int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, &v, &values_read);
+
+  if (values_read != 1) {
+    throw ParquetException("Failed to read value for column '" + node->name() + "'");
+  }
+}
+
+void StreamReader::read_(FixedLenByteArray& v) {
+  const auto& node{nodes_[column_index_]};
+  auto* const reader{
+      static_cast<FixedLenByteArrayReader*>(column_readers_[column_index_++].get())};
+  std::int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, &v, &values_read);
+
+  if (values_read != 1) {
+    throw ParquetException("Failed to read value for column '" + node->name() + "'");
+  }
+}
+
+void StreamReader::nextRow() {
+  column_index_ = 0;
+
+  if (column_readers_[0]->HasNext()) {
+    return;
+  }
+  nextRowGroup_();
+}
+
+void StreamReader::nextRowGroup_() {
+  // Find next none-empty row group
+  while (row_group_index_ < file_metadata_->num_row_groups()) {
+    row_group_reader_ = file_reader_->RowGroup(row_group_index_);
+    ++row_group_index_;
+
+    column_readers_.resize(file_metadata_->num_columns());
+
+    for (int i = 0; i < file_metadata_->num_columns(); ++i) {
+      column_readers_[i] = row_group_reader_->Column(i);
+    }
+    if (column_readers_[0]->HasNext()) {
+      return;
+    }
+  }
+  // No more row groups found.
+  eof_ = true;
+  file_reader_.reset();
+  file_metadata_.reset();
+  row_group_reader_.reset();
+  column_readers_.clear();
+  nodes_.clear();
+}
+
+void StreamReader::checkColumn_(const Type::type physical_type,
+                                const ConvertedType::type converted_type,
+                                const int length) {
+  if (static_cast<std::size_t>(column_index_) >= nodes_.size()) {
+    if (eof_) {
+      throw ParquetException("EOF reached");
+    }
+    throw ParquetException("Column index out-of-bounds.  Index " +
+                           std::to_string(column_index_) + " is invalid for " +
+                           std::to_string(nodes_.size()) + " columns");
+  }
+  const auto& node{nodes_[column_index_]};
```

Comment:
spell out type.

Reply:
I don't understand this comment which seems to be referring to the column index check.
There is no type associated with an invalid column index.

Reply:
I believe the suggestion is:
```suggestion
  const node_ptr_type& node{nodes_[column_index_]};
```

Reply:
Thanks for the clarification.
However, the whole point of the 'auto' keyword is to avoid spelling out types and make code easier to refactor.

Reply:
Please read https://google.github.io/styleguide/cppguide.html#Type_deduction for the reasoning.  There are pros and cons, however we do try to follow the style guide to avoid debates like this.

Reply:
The "style guide" says:
"For local variables, you can use type deduction to make the code clearer by eliminating type information that is obvious or irrelevant, so that the reader can focus on the meaningful parts of the code"
So why spell out a type which is irrelevant?

Reply:
The example in the style, guide give is eliding types on an iterator, but then recommenends the components of the iterator to spelled out types.  In this case without following the code, I'm not really sure what what node is (its an extra level of indirection).   There is some judgement here, so when a parquet committer does a final review they can give guidance.